### ISSUE TITLE
guidelines for DDP training in multinode environments

### ISSUE THREAD

**Issue by Akhil-CM** [webdataset/webdataset#414]
I'm trying to understand the proper pipeline for DDP training on multinode clusters with webdataset. There are some comments in [3] which I quote:
> 
    you may or may not want to split by note or worker
    you probably don't want resampled during validation
    you don't need shardshuffle when you are using resampled

in [1] I can see `nodesplitter=wds.split_by_node`.
**(Q1)** Does this mean the following from [the comment](https://github.com/webdataset/webdataset/issues/250#issuecomment-2010949877) in [4]:
> In many environments, we want every shard to be available on every node. If you are trying to split shards by node so that each node only sees a subset, you can use DataPipeline to do that.

**(Q2)** In case every shard is available on every node (by not setting `nodesplitter=wds.split_by_node`?), how do we avoid getting multiple duplicates of samples in an epoch?

**(Q3)** Why does `WebLoader` using `WebDataset` not need a  `DistributedSampler` whereas the indexed version `wids` needs `wids.DistributedChunkedSampler` in the corresponding `dataloader`

### Relevant info
***
1. [WebDataset + Distributed PyTorch Training](https://github.com/webdataset/webdataset/blob/main/examples/out/train-resnet50-multiray-wds.ipynb)
2. [WebIndexedDataset + Distributed PyTorch Training](https://github.com/webdataset/webdataset/blob/main/examples/out/train-resnet50-multiray-wids.ipynb)
3. https://github.com/webdataset/webdataset/issues/388
4. https://github.com/webdataset/webdataset/issues/250

**Comment by tmbdev** [comment]
PyTorch DataLoader operates in two fundamentally different modes:

- indexable datasets
- iterable datasets

It behaves very differently in those two modes. For single GPU training, the differences aren't all that apparent, but for multi-GPU training, they matter a great deal. In particular, PyTorch has no provisions for balancing samples across nodes when using iterable datasets. That's not a limitation of WebDataset, it's a limitation of PyTorch.

At the same time, the notion of "epoch" = "single pass through the dataset" is really pretty obsolete as well; it comes from the derivation of stochastic gradient descent as an approximation of gradient descent of the error function, but for most training these days, the "epoch" is not a useful unit of training anymore.

WebDataset gives you a choice of indexable datasets or iterable datasets. The recommended modes of operations are:

1. traditional epochs with single GPU training and iterable dataset
2. training using resampling and iterable datasets for single or multi GPU training (no epochs, just periodic reporting)
3. training using indexable datasets (wids) for multi GPU training if you want traditional epochs

The recommended way is really 2. With 2, you still have to decide whether all nodes sample from all shards or just a subset of shards.

The most familiar/traditional approach is probably 3.

What you can't easily do (though you can do with some effort) is use traditional epochs, iterable datasets, and multi-GPU/multi-node training; that's not a limitation of WebDataset, it's a limitation of PyTorch.

**Comment by ecolss** [comment]
@tmbdev very good advice!

> training using resampling and iterable datasets for single or multi GPU training (no epochs, just periodic reporting)
One thing I'm not very sure about is, let's say we're using iterable datasets, just like streaming data, then how does resampling works in this streamed data for DDP training?
